### PR TITLE
Changes for session and schema

### PR DESCRIPTION
**Pull request by mjok**
This need a discussion first.
Background liquibase cassandra uses outdated Simba JDBC driver; This project would be perfect fit for replacing it.

First thing, there is different concepts about catalog, You use cluster as catalog, as liquibase (simba) expects keyspace. And thus ailing on some checks:
getColumns <..>
if (catalog != null && !catalog.equals(this.connection.getCatalog()))

The solution could be either make this configurable, or suggest another solution.

Update also need review. Your return 0, with -1 "liquibase cassandra" would work out of the box. Yet again, this can be made configurable.

And passing a session would benefit. Actually I'd like to use the session from my application, there all the security params already configured, and there is no need to open additional session, as I have already one Idle.

note:
```
    private final TreeSet<String> hostListPrimary;
    private final TreeSet<String> hostListBackup;
```
seems to be unused variables

**sorry made a mess with previous pull request

### CODE REVIEW

**Comment by maximevw** [comment]
Thank you for taking the time to suggest improvements to this library.
I will answer your first comment below.

> This need a discussion first. Background liquibase cassandra uses outdated Simba JDBC driver; This project would be perfect fit for replacing it.
> 
> First thing, there is different concepts about catalog, You use cluster as catalog, as liquibase (simba) expects keyspace. And thus ailing on some checks: getColumns <..> if (catalog != null && !catalog.equals(this.connection.getCatalog()))
> 
> The solution could be either make this configurable, or suggest another solution.
> 
I perfectly understand your point. We kept the following correspondance of concepts as it existed in the library originally developed by `@adejanovski` (https://github.com/adejanovski/cassandra-jdbc-wrapper).
We think this is the most accurate correspondance regarding the definition of JDBC concepts.

| JDBC concept | Correspondance in Cassandra | Correspondance in a SQL database |
|-----------------|----------------------------------|----------------------------------------|
| Catalog           | Cluster                                     | Database or package                        |
| Schema          | Keyspace                                  | Schema                                             |
| Table              | Table                                         | Table                                                 |

So, if we consider that the catalog concept matches the Cassandra keyspace, what about the schema concept? Should we match it with the keyspace too? Why not... if we make it configurable to fit different needs like being compatible with Liquibase Cassandra. But I'm not sure it makes it clear to use it as the default implementation. I'll come back to it later.

> Update also need review. Your return 0, with -1 "liquibase cassandra" would work out of the box. Yet again, this can be made configurable.
> 
Even if we'll never be JDBC compliant because Cassandra is obviously not SQL 92 compliant 😉, we try to stick as more as possible to the [JDBC 4.3 API specifications](https://download.oracle.com/otn-pub/jcp/jdbc-4_3-mrel3-eval-spec/jdbc4.3-fr-spec.pdf?AuthParam=1647007575_e1a362066eef454ee5fc0e4b860ee87a).

So, regarding the method `executeUpdate()`, the specification mention (page 95):

> ## 13.1.2.2 Returning an Update Count
> In CODE EXAMPLE 13-5, the SQL statement being executed returns the number of rows affected by the update for SQL Data  Manipulation Language (DML) statements or 0 for SQL statements that return nothing.

So, we expect to return 0 since the driver doesn't return the number of updated rows.

But, once again, for specific needs, we could make it configurable. One solution would be to introduce a parameter `complianceMode=liquibase | default` (names can be discussed) in the JDBC URL indicating we want to deviate from the standard implementation to use a specific one. For example, if we want to be compliant with Liquibase Cassandra, we set the parameter value to `liquibase`. Then, instead of instantiating the default implementation classes (like `CassandraStatement`), it will used overridden classes (e.g. `LiquibaseCassandraStatement`) with specific behaviors. It's an idea, we can discuss it to avoid unecessary complexity.

> And passing a session would benefit. Actually I'd like to use the session from my application, there all the security params already configured, and there is no need to open additional session, as I have already one Idle.

I don't see any inconvenient to add such a constructor.

> note:
> 
> ```
>     private final TreeSet<String> hostListPrimary;
>     private final TreeSet<String> hostListBackup;
> ```
> 
> seems to be unused variables
> 
You're right, probably an oversight after a big refactoring :) We can remove them.

**Comment by maximevw** [comment]
Thank you for taking the time to suggest improvements to this library.
I will answer your first comment below.

> This need a discussion first. Background liquibase cassandra uses outdated Simba JDBC driver; This project would be perfect fit for replacing it.
> 
> First thing, there is different concepts about catalog, You use cluster as catalog, as liquibase (simba) expects keyspace. And thus ailing on some checks: getColumns <..> if (catalog != null && !catalog.equals(this.connection.getCatalog()))
> 
> The solution could be either make this configurable, or suggest another solution.
> 
I perfectly understand your point. We kept the following correspondance of concepts as it existed in the library originally developed by `@adejanovski` (https://github.com/adejanovski/cassandra-jdbc-wrapper).
We think this is the most accurate correspondance regarding the definition of JDBC concepts.

| JDBC concept | Correspondance in Cassandra | Correspondance in a SQL database |
|-----------------|----------------------------------|----------------------------------------|
| Catalog           | Cluster                                     | Database or package                        |
| Schema          | Keyspace                                  | Schema                                             |
| Table              | Table                                         | Table                                                 |

So, if we consider that the catalog concept matches the Cassandra keyspace, what about the schema concept? Should we match it with the keyspace too? Why not... if we make it configurable to fit different needs like being compatible with Liquibase Cassandra. But I'm not sure it makes it clear to use it as the default implementation. I'll come back to it later.

> Update also need review. Your return 0, with -1 "liquibase cassandra" would work out of the box. Yet again, this can be made configurable.
> 
Even if we'll never be JDBC compliant because Cassandra is obviously not SQL 92 compliant 😉, we try to stick as more as possible to the [JDBC 4.3 API specifications](https://download.oracle.com/otn-pub/jcp/jdbc-4_3-mrel3-eval-spec/jdbc4.3-fr-spec.pdf?AuthParam=1647007575_e1a362066eef454ee5fc0e4b860ee87a).

So, regarding the method `executeUpdate()`, the specification mention (page 95):

> ## 13.1.2.2 Returning an Update Count
> In CODE EXAMPLE 13-5, the SQL statement being executed returns the number of rows affected by the update for SQL Data  Manipulation Language (DML) statements or 0 for SQL statements that return nothing.

So, we expect to return 0 since the driver doesn't return the number of updated rows.

But, once again, for specific needs, we could make it configurable. One solution would be to introduce a parameter `complianceMode=liquibase | default` (names can be discussed) in the JDBC URL indicating we want to deviate from the standard implementation to use a specific one. For example, if we want to be compliant with Liquibase Cassandra, we set the parameter value to `liquibase`. Then, instead of instantiating the default implementation classes (like `CassandraStatement`), it will used overridden classes (e.g. `LiquibaseCassandraStatement`) with specific behaviors. It's an idea, we can discuss it to avoid unecessary complexity.

> And passing a session would benefit. Actually I'd like to use the session from my application, there all the security params already configured, and there is no need to open additional session, as I have already one Idle.

I don't see any inconvenient to add such a constructor.

> note:
> 
> ```
>     private final TreeSet<String> hostListPrimary;
>     private final TreeSet<String> hostListBackup;
> ```
> 
> seems to be unused variables
> 
You're right, probably an oversight after a big refactoring :) We can remove them.

**Comment by maximevw** [comment]
Please note I cleaned the unused code mentioned in your comment on the branch `release/next`.

**Comment by mjok** [comment]
Have no idea why build pipeline fails and
I wasn't able to run tests...

 java.lang.IllegalAccessError: superclass access check failed: class org.apache.cassandra.utils.JMXServerUtils$JmxRegistry (in unnamed module @0x3498ed) cannot access class sun.rmi.registry.RegistryImpl (in module java.rmi) because module java.rmi does not export sun.rmi.registry to unnamed module @0x3498ed

**Comment by maximevw** [comment]
Here is what I got in the build pipeline:
```
[INFO] Running com.ing.data.cassandra.jdbc.MetadataResultSetsUnitTest
[ERROR] Tests run: 6, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 3.227 s <<< FAILURE! - in com.ing.data.cassandra.jdbc.MetadataResultSetsUnitTest
[ERROR] givenStatement_whenMakeSchemas_returnExpectedResultSet  Time elapsed: 0.012 s  <<< FAILURE!
org.opentest4j.AssertionFailedError: expected: <embedded_test_cluster> but was: <test_keyspace>
	at com.ing.data.cassandra.jdbc.MetadataResultSetsUnitTest.givenStatement_whenMakeSchemas_returnExpectedResultSet(MetadataResultSetsUnitTest.java:89)

[ERROR] givenStatement_whenMakeCatalogs_returnExpectedResultSet  Time elapsed: 0.001 s  <<< FAILURE!
org.opentest4j.AssertionFailedError: expected: <embedded_test_cluster> but was: <test_keyspace>
	at com.ing.data.cassandra.jdbc.MetadataResultSetsUnitTest.givenStatement_whenMakeCatalogs_returnExpectedResultSet(MetadataResultSetsUnitTest.java:72)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   MetadataResultSetsUnitTest.givenStatement_whenMakeCatalogs_returnExpectedResultSet:72 expected: <embedded_test_cluster> but was: <test_keyspace>
[ERROR]   MetadataResultSetsUnitTest.givenStatement_whenMakeSchemas_returnExpectedResultSet:89 expected: <embedded_test_cluster> but was: <test_keyspace>
[INFO] 
[ERROR] Tests run: 182, Failures: 2, Errors: 0, Skipped: 0
```

I changed the configuration of the CI pipeline in Github actions to also build pull requests branches (I just updated the branch `release/next`).

I guess the error you mentioned comes from your local build. This error could happen if you run tests with JDK 9+ and not allowing illegal accesses. Try to build using JDK8 or tell me how you launched the tests.

**Comment by mjok** [comment]
Yes, you're right. JDK 1.8 works fine. 
And had to shutdown Cassandra I was running locally. My bad.

Glad it worked, find some quirkies I've lost at first.

**Comment by maximevw** [comment]
Do you see the pipeline logs here: https://github.com/ing-bank/cassandra-jdbc-wrapper/actions/workflows/ci-workflow.yml ?

For Azure pipeline, I restarted it, it failed when it checked-out the repo from Github... and now it succeeded.

**Comment by mjok** [comment]
Yes, I can.

File: src/main/java/com/ing/data/cassandra/jdbc/Utils.java [hunk 1]
```diff
@@ -135,6 +135,7 @@
     public static final String TAG_CLOUD_SECURE_CONNECT_BUNDLE = "secureConnectBundle";
     public static final String TAG_CONFIG_FILE = "configFile";
     public static final String TAG_REQUEST_TIMEOUT = "requestTimeout";
+    public static final String TAG_COMPLIANCE_MODE = "complianceMode";
```
Comment (maximevw) [inline src/main/java/com/ing/data/cassandra/jdbc/Utils.java:138]: You also have to create a cosntant "KEY_COMPLIANCE_MODE" (see above in the code) with a full lower case name to take it into account in the JDBC URL as query parameter, then map it to the corresponding tag after the line 309 in this file (method `parseURL`). This is probably not very optimized but refactoring this part, inherited from the original project, is not currently the priority.

File: src/main/java/com/ing/data/cassandra/jdbc/CassandraConnection.java [hunk 2]
```diff
@@ -187,7 +227,9 @@ public void clearWarnings() throws SQLException {
 
     @Override
     public void close() throws SQLException {
-        this.sessionHolder.release();
+        if (sessionHolder!= null) {
```
Comment (maximevw) [inline src/main/java/com/ing/data/cassandra/jdbc/CassandraConnection.java:230]: Please surround '!=' with spaces.

File: src/main/java/com/ing/data/cassandra/jdbc/CassandraConnection.java [hunk 3]
```diff
@@ -166,6 +183,29 @@ public CassandraConnection(final SessionHolder sessionHolder) throws SQLExceptio
         });
     }
 
+    public CassandraConnection(Session cSession, String currentKeyspace, ConsistencyLevel defaultConsistencyLevel, boolean debugMode) {
```
Comment (maximevw) [inline src/main/java/com/ing/data/cassandra/jdbc/CassandraConnection.java:186]: To be compliant with our Checkstyle rules, please mark each parameter as final:
```
[WARN] /home/runner/work/cassandra-jdbc-wrapper/cassandra-jdbc-wrapper/src/main/java/com/ing/data/cassandra/jdbc/CassandraConnection.java:186:32: Parameter cSession should be final. [FinalParameters]
[WARN] /home/runner/work/cassandra-jdbc-wrapper/cassandra-jdbc-wrapper/src/main/java/com/ing/data/cassandra/jdbc/CassandraConnection.java:186:50: Parameter currentKeyspace should be final. [FinalParameters]
[WARN] /home/runner/work/cassandra-jdbc-wrapper/cassandra-jdbc-wrapper/src/main/java/com/ing/data/cassandra/jdbc/CassandraConnection.java:186:74: Parameter defaultConsistencyLevel should be final. [FinalParameters]
[WARN] /home/runner/work/cassandra-jdbc-wrapper/cassandra-jdbc-wrapper/src/main/java/com/ing/data/cassandra/jdbc/CassandraConnection.java:186:116: Parameter debugMode should be final. [FinalParameters]
```
(you can find this in the build logs)

File: src/main/java/com/ing/data/cassandra/jdbc/CassandraConnection.java [hunk 4]
```diff
@@ -166,6 +183,29 @@ public CassandraConnection(final SessionHolder sessionHolder) throws SQLExceptio
...
+        this.sessionHolder = null;
+        this.connectionProperties = new Properties();
+        this.optionSet = optionSet == null ? lookupOptionSet(null) : optionSet;
```
Comment (maximevw) [inline src/main/java/com/ing/data/cassandra/jdbc/CassandraConnection.java:189]: Even it's less verbose, please don't use inline conditionals as per our Checkstyle rules:
```
[WARN] /home/runner/work/cassandra-jdbc-wrapper/cassandra-jdbc-wrapper/src/main/java/com/ing/data/cassandra/jdbc/CassandraConnection.java:189:44: Avoid inline conditionals. [AvoidInlineConditionals]
```

File: src/main/java/com/ing/data/cassandra/jdbc/CassandraConnection.java [hunk 5]
```diff
@@ -504,4 +533,25 @@ public String toString() {
         throw new SQLFeatureNotSupportedException(String.format(NO_INTERFACE, iface.getSimpleName()));
     }
 
+
+    public OptionSet getOptionSet() {
+        return optionSet;
+    }
+
+    private OptionSet lookupOptionSet(String property) {
+        ServiceLoader<OptionSet> loader = ServiceLoader
```
Comment (maximevw) [inline src/main/java/com/ing/data/cassandra/jdbc/CassandraConnection.java:542]: The local variables and method parameters should be declared final as per our Checkstyle rules:
```
[WARN] /home/runner/work/cassandra-jdbc-wrapper/cassandra-jdbc-wrapper/src/main/java/com/ing/data/cassandra/jdbc/CassandraConnection.java:541:39: Parameter property should be final. [FinalParameters]
[WARN] /home/runner/work/cassandra-jdbc-wrapper/cassandra-jdbc-wrapper/src/main/java/com/ing/data/cassandra/jdbc/CassandraConnection.java:542:34: Variable 'loader' should be declared final. [FinalLocalVariable]
[WARN] /home/runner/work/cassandra-jdbc-wrapper/cassandra-jdbc-wrapper/src/main/java/com/ing/data/cassandra/jdbc/CassandraConnection.java:544:29: Variable 'iterator' should be declared final. [FinalLocalVariable]
[WARN] /home/runner/work/cassandra-jdbc-wrapper/cassandra-jdbc-wrapper/src/main/java/com/ing/data/cassandra/jdbc/CassandraConnection.java:546:23: Variable 'optionSet' should be declared final. [FinalLocalVariable]
[WARN] /home/runner/work/cassandra-jdbc-wrapper/cassandra-jdbc-wrapper/src/main/java/com/ing/data/cassandra/jdbc/CassandraConnection.java:552:19: Variable 'optionSet' should be declared final. [FinalLocalVariable]
```

**Review by maximevw — CHANGES_REQUESTED** [review]
To be compliant with our contributing policy, please:
- add license headers in the new Java files
- add Javadoc on the public methods and constructors (you can find Checkstyle findings in the build logs)
- add unit tests for the new code each time it is possible

Thank you

**Pull request merged**